[material-ui][InputBase] Text cursor jumps several lines up when inserting text fragment with new line at the end

yermartee opened the pull request:
Fixes https://github.com/mui/material-ui/issues/45148
Fixes https://github.com/mui/material-ui/issues/45277

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

DiegoAndai reviewed (commented):
Hey, @yermartee, thanks for working on this.

May I ask you to explain the cause of the issue and how this PR solves it?

yermartee commented:
Hi @DiegoAndai, we initially thought it was a MUI bug, but it appears to be a browser issue related to textareas. (This textarea issue appears to be browser-specific, affecting multiple implementations with the notable exception of Mozilla.)

The stringified value of the textarea input is "abc def abc def abc def\nabc def abc def abc def\nabc def abc def abc def\n". I suspect this newline character at the end is causing the caret jumping issue.

Adding a space before the final newline (resulting in "abc def abc def abc def\nabc def abc def abc def\nabc def abc def abc def \n") seems to work as a workaround. However, this is not a stable solution and I am continuing to investigate.

I've attached a screen recording from Chrome, showing the issue in a plain HTML textarea. 
Please advise on whether MUI should address this common textarea issue. If the decision is to proceed, please hold off on reviewing this pull request until I can deliver a stable solution.

https://github.com/user-attachments/assets/a59c576c-16d8-4b8f-af68-ab859c27d9fd

DiegoAndai commented:
Thanks for the explanation @yermartee. If we can solve it on our side, that would be good. Let me know if you find a way 👍🏼 

Also, please look at https://github.com/mui/material-ui/issues/45277, as it might be related.

yermartee commented:
Hi @DiegoAndai the previous solution wasn't reliable, so I've changed it. 

Now, instead of messing with the new state or other tricky stuff, I'm using a simple trick to keep your cursor at the end of the text box, even when you hit Enter (and add a new line). 

This code checks if you typed a new line and if your cursor's already at the end, then puts it back at the very end.

I've checked the https://github.com/mui/material-ui/issues/45277 yes it's related.
The attached screen recordings demonstrate that I have addressed these issues.

Fixes https://github.com/mui/material-ui/issues/45148 
https://github.com/user-attachments/assets/0ef0af30-5188-47d4-a43b-71e82aaff1a2

Fixes https://github.com/mui/material-ui/issues/45277
https://github.com/user-attachments/assets/b29d1530-394f-4137-b3c0-9ace73ad21c0

Please review, Thanks in advance.

DiegoAndai reviewed (commented):
The solution makes sense @yermartee 👌🏼 

We should add a test for this, can you give it a try?

yermartee commented:
Hey @DiegoAndai I've added a test for the input caret jumping issue.  Please review it and let me know if there are any ways to improve it. Thanks.

DiegoAndai reviewed (commented):
Hey @yermartee! Thanks for adding the test.

Just one question about it: If you remove the fix, does the test fail?

yermartee commented:
Hey @DiegoAndai, thank you for feedback.

As I mentioned earlier, it's a browser-specific bug that we're fixing for the MUI TextareaAutosize component.

As I understand, we're using React Testing Library with JSDom, which are isolated environments and cannot perfectly mimic Chrome's browser behavior.

I've tried various ways to write tests that better approximate Chrome's behavior, but I haven't been able to find a test case that can catch this bug in JSDom. 

Therefore, after removing the current solution, the test will no longer fail. What should we do in this case? Please advise. Thanks.

DiegoAndai commented:
@yermartee we also have broswer tests you can run with `pnpm test:karma`, this runs in Chrome and other browsers. Let me know if you are able to reproduce the behavior using that.

yermartee commented:
Hi @DiegoAndai thank you for the guidance. 

I've run the Karma test and enhanced the test cases to closely match our real-world scenario. 

However, I'm still unable to reproduce the caret jumping behavior in headless Chrome that we observe in Real Chrome browser usage.

Key Observations:

<!--StartFragment-->
Operation | Real Chrome  | Headless Chrome | Test Result
-- | -- | -- | --
Initial Caret Position | selectionStart: 0 | selectionStart: 0 | -
After 1st Paste | 24 | 24 | Pass
After 2nd Paste | 48 | 48 | Pass
After 3rd Paste | 0 | 72 | Fails on Real Chrome / Passes on Headless Chrome
<!--EndFragment-->

Is it possible to write Playwright tests instead of RTL and run them to real chrome?

DiegoAndai commented:
@yermartee thanks for trying! I think we have one final option, adding a regression test, which uses playwright. Here's an example for the Autocomplete component: https://github.com/mui/material-ui/blob/master/test/regressions/index.test.js#L138. I think we could do something similar.

And here are the commands to run these tests: https://github.com/mui/material-ui/blob/master/test/regressions/README.md#commands

Let me know if this works 😊

yermartee commented:
Hey @DiegoAndai Thanks for helping. I've added a regression test for TextareaAutosize, and it covers the test case.

yermartee commented:
Hey @DiegoAndai  Thanks for helping. I've added a regression test for TextareaAutosize, and it covers the test case.

Also, please advise: should we take screenshots or just add a check for input caret jumping?

DiegoAndai commented on test/regressions/fixtures/Textarea/TextareaAutosize.js:
```diff
@@ -0,0 +1,25 @@
+      minRows={4}
+      onChange={handleChange}
+      value={value}
+      slotProps={{
+        input: {
+          role: 'textbox',
+        },
+      }}
```
Is this required?

yermartee replied:
I think yes, we need to target the textbox by its role. We're using it in page.getByRole('textbox').

DiegoAndai replied:
Right, but I think the `textarea` element should have the `textbox` role by default, doesn't it?

yermartee replied:
Right, but in that case, I cannot get access to the input. Instead, I am getting access to the div element, which is the parent of the input.

DiegoAndai replied:
Ah, I see. In that case I think we should do

```
      slotProps={{
        input: {
          'data-testid': 'input',
        },
      }}
```

And use `page.getByTestId('input')`

Would that work?

yermartee replied:
It works well and has been applied.

DiegoAndai commented on test/regressions/index.test.js:
```diff
@@ -158,6 +158,38 @@ async function main() {
+        await page.keyboard.press(pasteCommand);
+        await takeScreenshot({ testcase, route: '/regression-Textarea/TextareaAutosize1' });
+
+        await page.waitForTimeout(500);
```
Are the timeouts required?

yermartee replied:
Since we can skip the screenshots part, maybe we can go without timeouts. I need to check it.

yermartee replied:
Removed timeouts and tested; it works well. Thanks a lot for the help! @DiegoAndai

DiegoAndai reviewed (commented):
Thanks @yermartee!

> should we take screenshots or just add a check for input caret jumping?

Is the check for input caret enough for the test? If it is, then we can skip the screenshots.

DiegoAndai commented:
> However, the regression tests are still automatically capturing screenshots during fixture rendering

That's ok 👍🏼

DiegoAndai commented:
Thanks for the contribution @yermartee!

DiegoAndai commented:
cc @mnajdova @siriwatknp I merged this community contribution fix. I will be out for some weeks, so just a ping so someone is aware in case we have a regression while I'm not here.
